src/Microsoft.Diagnostics.Monitoring.EventPipe/Triggers/Pipelines/EventPipeTriggerPipeline.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.EventPipe.Triggers.Pipelines
+{
+    /// <summary>
+    /// Starts an event pipe session using the specified configuration and 
+    /// </summary>
+    /// <typeparam name="TSettings">The settings type of the trace event trigger.</typeparam>
+    internal sealed class EventPipeTriggerPipeline<TSettings> :
+        EventSourcePipeline<EventPipeTriggerPipelineSettings<TSettings>>
+    {
+        // The callback as provided to the pipeline. Invoked when the trigger condition is satisfied.
+        // The trigger condition may be satisfied more than once (thus invoking the callback more than
+        // once) over the lifetime of the pipeline, depending on the implementation of the trigger.
+        private readonly Action<TraceEvent> _callback;
+
+        /// <summary>
+        /// The pipeline used to monitor the trace event source from the event pipe using the trigger
+        /// specified in the settings of the current pipeline.
+        /// </summary>
+        private TraceEventTriggerPipeline _pipeline;
+
+        // The trigger implementation used to detect a condition in the trace event source.
+        private ITraceEventTrigger _trigger;
```

Comment:
I know you discuss this in your comments to some extent, but would it be reasonable to allow multiple ITraceEventTrigger objects here now? Each trigger object could check if an event satisfies the trigger condition. The provider configuration is separate anyway so even we don't have code that merges provider configurations, this could still work. You would still need to merge the provider map from each trigger and perhaps only send events to triggers whose provider map matches the current event.

Reply:
At this level (EventPipeTriggerPipeline), I would think that the ability to add multiple triggers would have to come at the same time as merging configurations together. What I'm imagining is something like:

```
EventPipeTriggerPipelineSettings pipelineSettings = new();
pipelineSettings.EventProviderMerger.Add(MicrosoftExtensionsLoggingMerger.Instance);
pipelineSettings.EventProviderMerger.Add(MicrosoftDiagnosticsDiagnosticSourceMerger.Instance);

pipelineSettings.TryAddTrigger(trigger1Factory, trigger1Settings, trigger1Configuration);
pipelineSettings.TryAddTrigger(trigger2Factory, trigger2Settings, trigger2Configuration);
pipelineSettings.TryAddTrigger(trigger3Factory, trigger3Settings, trigger3Configuration);
...
```

I would want to reject adding a new trigger if the required configuration of that trigger cannot be merged with the existing triggers on the settings. For example:
- two event counter triggers are added for the same event provider but have different intervals; these cannot be merged because the EventCounters infrastructure only expects one interval on each provider for an event pipe session (I'm assuming this; I'll go double check that this is the case).
- two diagnostic source based triggers are added and both assume they can use Activity1Start and Activity1Stop. Unless they are using the same exact event and have payload properties that can be cleanly merged, then they cannot run in the same event pipe session. Just forgetting about the configuration for a moment, both of these triggers cannot be run together whatsoever because the Activity1Start and Activity1Stop events will be conflated.
- two logging based triggers may specify the same logger categories at different levels. While it's probably not the end of the world to forward events from a lower log level (e.g. Debug) to the trigger that requested more restrictive level (e.g. Warning), that other trigger might not expect that and behavior badly.

Rejecting triggers with incompatible configurations allows the settings to be successful when the pipeline is started and allows the caller to understand that specific trigger configurations are incompatible. The caller then has more information as to why a certain trigger just cannot be satisfied even before starting the pipeline. The caller can take appropriate action when this happens, e.g. aggregate the rejected triggers into another pipeline instance, if necessary.